[STORM-1230] port backtype.storm.process-simulator to java.

zhuoliu opened the pull request:
From storm-core/src/clj/org/apache/storm/process-simulator.clj to storm-core/src/jvm/org/apache/storm/ProcessSimulator.java

abhishekagarwal87 commented on storm-core/src/jvm/org/apache/storm/ProcessSimulator.java:
```diff
@@ -0,0 +1,89 @@
+            if (shutdownHandle != null) {
+                shutdownHandle.shutdown();
+            }
+            processMap.remove(pid);
```
I think this should be moved inside the if block and before the shutdown call. Once process is marked for shutdown, it should be removed from map (similar to clojure code) 
``` 
Shutdownable shutdownHandle = processMap.remove(pid)
```

zhuoliu replied:
Here I would prefer to keep it unchanged since It would be possible that <pid, null> be wrongly put in the map by registerPorcess. Place the "remove" outside of the "if" would help and not hurt, also this translation conforms the original clj code logic.

abhishekagarwal87 replied:
I meant it something such as follows 
```
Shutdownable shutdownHandle = processMap.remove(pid)
if (shutdownHandle != null) {
                shutdownHandle.shutdown();
            }
```
Sorry for the confusion.

abhishekagarwal87 commented on storm-core/src/jvm/org/apache/storm/ProcessSimulator.java:
```diff
@@ -0,0 +1,89 @@
+                shutdownHandle.shutdown();
+            }
+            processMap.remove(pid);
+            LOG.info("Successfully killing process " + pid);
```
typo: killed.

abhishekagarwal87 replied:
and should this be logged only if pid is present?

abhishekagarwal87 commented on storm-core/src/jvm/org/apache/storm/ProcessSimulator.java:
```diff
@@ -0,0 +1,89 @@
+
+public class ProcessSimulator {
+    private static Logger LOG = LoggerFactory.getLogger(ProcessSimulator.class);
+    protected static Object lock = new Object();
```
why are these declared protected?

abhishekagarwal87 commented on storm-core/src/jvm/org/apache/storm/ProcessSimulator.java:
```diff
@@ -0,0 +1,89 @@
+     * @param pid
+     * @return
+     */
+    protected static Shutdownable getProcessHandle(String pid) {
```
This function is only used in this class. It can be inlined.

abhishekagarwal87 commented on storm-core/src/jvm/org/apache/storm/ProcessSimulator.java:
```diff
@@ -0,0 +1,89 @@
+        for (String pid : pids) {
+            killProcess(pid);
+        }
+        LOG.info("Successfully kill all processes");
```
typo: killed. You can add a begin info message as well.

abhishekagarwal87 commented on storm-core/src/jvm/org/apache/storm/ProcessSimulator.java:
```diff
@@ -0,0 +1,89 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class ProcessSimulator {
```
Also ProcessSimulator name is misleading judging from how it is being used in supervisor. ProcessManager sounds a better name to me.

abhishekagarwal87 replied:
Ahh. Ignore that. code is doing something else.

revans2 replied:
Can we put in some javadocs to explain that the ProcessSimulator keeps track of Shutdownable objects in place of actual processes.

Also can we file a JIRA to replace Shutdownable with AutoClosable?

revans2 commented on storm-core/src/jvm/org/apache/storm/ProcessSimulator.java:
```diff
@@ -0,0 +1,80 @@
+        for (String pid : pids) {
+            killProcess(pid);
+        }
+        LOG.info("Successfully killed all processes");
```
We have extra logging here.  In the original code the only log message was once for each process that was "Killed".

revans2 commented:
Only two very minor comments, other even without them I am a +1

zhuoliu commented:
Hi @revans2, really appreciate your comments! I addressed your comment in [Storm-1563](https://issues.apache.org/jira/browse/STORM-1563) to change all the Shutdownable to be AutoClosable in storm.

redsanket commented:
upmerge

zhuoliu commented:
@redsanket thanks, upmerged.

redsanket commented on storm-core/src/jvm/org/apache/storm/ProcessSimulator.java:
```diff
@@ -0,0 +1,82 @@
+    }
+
+    /**
+     * kill all processes
```
capital "Kill" just for consistency

redsanket commented:
Very minor comment, not significant LGTM +1
